### PR TITLE
Feature fix os pathlib

### PR DESCRIPTION
**Pull request by gosia-malgosia**

### CODE REVIEW

File: imageatm/components/cloud.py [hunk 1]
```diff
@@ -73,37 +78,33 @@ def _set_remote_dirs(self):
...
-            )
+            self.s3_job_dir = self.s3_bucket / 'job_dirs' / self.job_dir.name
 
     def _sync_local_s3(self):
         self.logger.info('Syncing files local <> s3...')
         self._set_s3_dirs()
 
         # only sync if image dir is local dir
-        if 's3://' not in self.image_dir:
+        if isinstance(self.image_dir, Path):
```
Comment (clennan) [inline imageatm/components/cloud.py]: Previous code is more specific in my opinion. 
```suggestion
        if 's3://' not in str(self.image_dir):
```

File: imageatm/components/cloud.py [hunk 2]
```diff
@@ -73,37 +78,33 @@ def _set_remote_dirs(self):
...
             cmd = 'aws s3 sync --quiet --exclude logs {} {}'.format(
                 self.image_dir, self.s3_image_dir
             )
             run_cmd(cmd, logger=self.logger)
 
         # only sync if job dir is local dir
-        if 's3://' not in self.job_dir:
+        if isinstance(self.job_dir, Path):
```
Comment (clennan) [inline imageatm/components/cloud.py]: same as above

File: imageatm/components/cloud.py [hunk 3]
```diff
@@ -112,7 +113,7 @@ def _sync_s3_local(self):
         self._set_s3_dirs()
 
         # only sync if job dir is local dir
-        if 's3://' not in self.job_dir:
+        if isinstance(self.job_dir, Path):
```
Comment (clennan) [inline imageatm/components/cloud.py]: same as above

File: imageatm/components/data_prep.py [hunk 4]
```diff
@@ -77,7 +76,7 @@ def __init__(
         """
         self.job_dir = Path(job_dir).resolve()
         if not self.job_dir.exists():
```
Comment (clennan) [inline imageatm/components/data_prep.py]: We could make use of `exist_ok`
```suggestion
            self.job_dir.mkdir(parents=True, exist_ok=True)
```

File: imageatm/components/data_prep.py [hunk 5]
```diff
@@ -307,8 +306,8 @@ def _split_samples(self):
     def _resize_images(self, resize_image_mp: Callable = resize_image_mp):
         self.logger.info('\n****** Resizing images ******\n')
         new_image_dir = '_'.join([str(self.image_dir), 'resized'])
-        if not os.path.exists(new_image_dir):
-            os.makedirs(new_image_dir)
+        if not Path(new_image_dir).resolve().exists():
```
Comment (clennan) [inline imageatm/components/data_prep.py]: use `.mkdir(parents=True, exist_ok=True)`

File: imageatm/components/evaluation.py [hunk 6]
```diff
@@ -91,19 +90,19 @@ def _load_best_model(self):
...
     def _create_evaluation_dir(self):
         """Creates evaluation dir for reporting."""
         if self.save_plots:
-            evaluation_dir_name = os.path.basename(self.best_model_file).split('.hdf5')[0]
+            evaluation_dir_name = self.best_model_file.name.split('.hdf5')[0]
             self.evaluation_dir = self.job_dir / 'evaluation_{}'.format(evaluation_dir_name)
 
             if not self.evaluation_dir.exists():
-                os.makedirs(self.evaluation_dir)
+                self.evaluation_dir.mkdir(parents=True)
```
Comment (clennan) [inline imageatm/components/evaluation.py]: use .mkdir(parents=True, exist_ok=True)

File: imageatm/components/training.py [hunk 7]
```diff
@@ -119,14 +119,14 @@ def _fit_model(self):
...
-        # tensorboard = TensorBoardBatch(log_dir=os.path.join(job_dir, 'logs'))
+        # tensorboard = TensorBoardBatch(log_dir=Path(job_dir).resolve() / 'logs')
 
         model_save_name = (
             'model_' + self.base_model_name.lower() + '_{epoch:02d}_{val_acc:.3f}.hdf5'
         )
         model_dir = self.job_dir / 'models'
         if not model_dir.is_dir():
-            os.makedirs(model_dir)
+            model_dir.mkdir(parents=True)
```
Comment (clennan) [inline imageatm/components/training.py]: use .mkdir(parents=True, exist_ok=True)

File: tests/utils/test_tf_keras.py [hunk 8]
```diff
@@ -17,10 +17,10 @@
 train_samples = 20
 test_samples = 20
 
-TEST_DIR = 'tests/data/test_callbacks/'
+TEST_DIR = Path('tests/data/test_callbacks/').resolve()
 
-if not os.path.exists(TEST_DIR):
-    os.makedirs(TEST_DIR)
+if not TEST_DIR.exists():
+    TEST_DIR.mkdir(parents=True)
```
Comment (clennan) [inline tests/utils/test_tf_keras.py]: use .mkdir(parents=True, exist_ok=True)

File: setup.py [hunk 9]
```diff
@@ -34,6 +34,7 @@
         'scikit-learn',
         'scipy',
         'tqdm',
+        'urlpath',
```
Comment (datitran) [inline setup.py]: What was the reason to use urlpath? Urlpath has only 25 stars and it's not really maintained.

Comment (gosia-malgosia) [reply setup.py]: pathlib is not suitable for urls (s3://) so I had to find a different library for handling urls. 
I changed to yarl library: https://github.com/aio-libs/yarl (387 stars)
Hopefully in the future this issue will be maintained within pathlib and we will have an easy way of switching URL to PosixPath

Whi I didn't go with urllib:
The built-in urllib.parse is lacking the functionality we need- we would have to convert the urls to strings and this means the methods that can take either path or url will only use string transformations. So for example it will be impossible to use pathlib's `image_dir.name` to extract the basename. Instead we will have to write `str(image_dir).split('/')[-1]` which to me makes the entire usage or pathlib redundant. The advantage of the urlpath is that it stores urls as urls while providing similar methods as in pathlib (so for example both the URL and Path objects have a method `name`, you can also join further levels of the path simply by adding slash+string: `base / next-level / image_dir` instead of `'/'.join((base, next-level, str(image_dir).split('/')[-1]))`
